DCOS-21752 DOCUMENTATION describes the mediator api

Fabs opened the pull request:
Presents an example of the mediator API and some documentation.

An implementation of the Settings > Package Repository page using this api is on the go here.

This documentation will be merged into `data-layer`. And once we are ok with all the pieces we will merge this integration in master

DanielMSchmidt commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,79 @@
+// The data composition bellow is using only streams, 
+// IF we had graphql, it could look like this
+/*
+  const query = "
```
Can you maybe extract the graphql example into a second example, it's kinda hard to comprehend this way. For me everything here is new and I need to sort out what happens, so thinking how it could be with graphql on thr side is too much for me 😇

Fabs replied:
I agree that that makes reading linearly harder. Will think about something. Thanks.

DanielMSchmidt replied:
a second complete example would be cool, I think that is the easiest to comprehend

Fabs replied:
Ok

Fabs commented:
@DanielMSchmidt the milestone for the data-layer is a consistent API + reference implementation + documentation. Those things need to come together in pieces first, and then we can have something for master. 

One way to see it as the data-service being a package, there is no PR to updated the version of the package on master, until there is a release of the package. The feature branch is that v0.1 release in progress. We don't wanna have to repos at this point for velocity reasons.

Fabs commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,79 @@
+This package is to be extracted to dcos-ui/data-service.
+
+#API v1
```
Thanks for that, my editor did not complained about it :/

Fabs commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,79 @@
+import { renderComponents } from "data-service"
+
+// The data composition bellow is using only streams, 
+// IF we had graphql, it could look like this
```
I actually put the big If on purpose, but you are right :D

DanielMSchmidt replied:
okay, thats fine

Fabs commented:
@DanielMSchmidt. I corrected the mistakes and added the second example. I also changed the renderComponents to the recompose version we are using. Please feel free to review again and challenge if you think it is not good or if you are still confused.

DanielMSchmidt commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,113 @@
+  graphql(launchMutation, shuttleSchema).mutate({ name, launchDate });
+}
+
+const components$ = content$.map((data) => 
```
How does the data get passed into the graphql query? I mean $location needs be specified for the query to be resolvable, right?

Fabs replied:
At this point of the writing that is not clear. I still have to figure out the details on how we are going to use graphl which will be the rest of the work I am doing. Notice that the location is hardcoded on the non-graphql example anyway. 

To make this example whole, I can remove `$location`.

Fabs replied:
I solved it in a different way, by adding an idea about how the parameters could be.

DanielMSchmidt commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,113 @@
+  }
+`;
+
+// This would be a graphql mutation
```
Does this comment make sense here?

Fabs replied:
Not anymore, no.

Fabs commented:
@DanielMSchmidt is there still any doubt, or any comment to address here?

juliangieseke commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,112 @@
+  launchShuttle({name, launchDate: new Date()})
+}
+
+const components$ = content$.map((data) => 
```
Is there any possibility to wrap this in a JSX-Component?

juliangieseke commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,112 @@
+
+const launch = (name) => {
+  const launchDate = new Date();
+  graphql(launchMutation, shuttleSchema).mutate({ name, launchDate });
```
shouldnt even `.mutate` return something "to work with" in order to be able to notify the component about changed data from somewhere else? how to we match this to a stream we're listening to?

Fabs replied:
I will stop providing an api for mutations as I have no idea how it is going to work for now.

Fabs commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,112 @@
+                                  <Space shuttles={data.launched} />
+                                  <LaunchButton onLaunch={launch} />
+                                </World>)
+                            .startWith(<LoadingUI />)                                
```
The pattern is to convert a stream of data into a stream of components. If there is one component handling loading and data, then the `startWith` move to the data part, and yields some information that the query started as first value `{isLoading: true}` for instance.

Same for the error, if there is a single component handling it, the catch move to the data part, and emits data when error happens `{isError: true, message: "blabla"`. 

I prefer the way I exemplified, but it is a question of choice.

Fabs commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,112 @@
+                                  <LaunchButton onLaunch={launch} />
+                                </World>)
+                            .startWith(<LoadingUI />)                                
+                            .catch((err) => <ErrorMessageUI err>));
```
see above :D

Fabs commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,112 @@
+// See bellow for the equivalent graphql.
+const launched$ = shuttles$.filter((sht) => sht.location === "space");
+const nextLaunches$ = shuttles$.filter((sht) => sht.location === "earth");
+const content$ = launched$.combineLatest(nextLaunches$,
```
I am replacing to a better example.

Fabs commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,112 @@
+  launchShuttle({name, launchDate: new Date()})
+}
+
+const components$ = content$.map((data) => 
```
You are right. The example is just not very good.

Fabs commented:
@juliangieseke I had a session with julian and I made some comments on his very valid questions.

I provided also a better working example now :D. And at this point in time removed everything about mutations. We will have then once I work on the api (this week yet). But they are are scoped also in a different ticket.

Fabs commented:
Can you look again @juliangieseke and @DanielMSchmidt.

Fabs commented:
Also, this can't fail right? :D

juliangieseke commented:
Did you rebase this lately? Its an integration test which failed three times 🤷‍♂️

Fabs commented:
it is just a README file, CI is broken, I would be in favor to get a second approve and merge even with CI failing.

nLight commented on packages/data-service/README.md:
````diff
@@ -0,0 +1,147 @@
+```js
+  /*
+   * Receives a stream of React.Component and outputs a single components that 
+   * renders the most recevent event from the Observable.
````
🔤 the most recent

Fabs replied:
✅

nLight commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+import {
+  createEventHandler,
+  componentFromStream
+} from "#PACKAGES/data-service/recomposeRx";
```
should be possible just `import * from "data-service/recomposeRx";` We had this alias in webpack so that `./packages` act as `node_modules`

juliangieseke replied:
I actually like being explicit here 🤔

Fabs replied:
#PACKAGES would be dcos-ui specific, and this package is going to be called data-service in the future, and move to dcos-labs, so actually "#PACKAGES" only make change if you know dcos-ui.

nLight commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+import { fetchRepositories } from "#PLUGINS/catalog/src/js/repositories/repositoriesStream";
+
+// The graphql schema and resolvers for those streams;
+import { typeDefs, resolvers } from "#SRC/js/model/repositoryModel";
```
I know it's an example but this belong to `"#PLUGINS/catalog..."`

Fabs replied:
You are right. On the actual implementation, I moved that on Friday. think that does not have any impact on the example.

nLight commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+  query: fetchRepositories(),
+  filter: searchTerm$
+})
+  .combineLatest(searchTerm$, (result, searchTerm) => {
```
Why do we need passing `searchTerm$` twice?

Fabs replied:
The first is passed as an argument to the query so it is used as a filter by graphql. You are filtering the PackageRepository mode, which has the fileds (name, priority and uri). But you also need to show the search term on the UI. Since the searchTerm is not a coming from the "data", but from the stream on `createEventHandler`, you have to combine it again.

The semantic is, combine the data stream with the events from my components and pass it down as props.

nLight commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+    );
+  })
+  // The first component is the loading
+  .startWith(<RepositoriesTabUI.Loading />)
```
Name is confusing 🤔

juliangieseke commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+  // The first component is the loading
+  .startWith(<RepositoriesTabUI.Loading />)
+  // If anything goes wrong, we render an error component
+  .catch(err => Observable.of(<RepositoriesTabUI.Error err={err} />));
```
again, short info what is catched here? (not errors from `<RepositoriesTabUI>` but stuff from the stream)

Fabs replied:
ok

juliangieseke commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+  })
+  // The first component is the loading
+  .startWith(<RepositoriesTabUI.Loading />)
+  // If anything goes wrong, we render an error component
```
see note above 👍

juliangieseke commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+
+### Mutations
+
+TBD
```
Ah ok 👍

Fabs commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+
+// Using the data layer
+
+// 1. We first make an schema out of the resolvers and typeDefinitions
```
Agree.

Fabs commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+  resolvers
+});
+
+// 2. We declare which data our component needs in graphql (data down)
```
It is assumed you know graphql and rxjs. Otherwise I would be written a bunch of stuff the internet does much better. I added the schema and the resolvers as asked on the previous comment, that should be enough to understand given those assumptions.

Fabs commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+
+// 3. We create an event handler as a stream to deal with search (events up)
+const { handler: onSearch, stream: searchEvent$ } = createEventHandler();
+const searchTerm$ = searchEvent$.startWith("");
```
done.

Fabs commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+
+## Mediator Pattern
+
+The [mediator pattern](https://en.wikipedia.org/wiki/Mediator_pattern) is used here
```
I believe the pattern is an idea, not the implementation. It can manifest in many ways. We use the idea to come out with our solution.

Fabs commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+
+// 4. We compose the result with the query, and massage it down the pipe until
+// we get the props for our component.
+// The idea is from an Observable of data and events, compose a stream o React.Components
```
thanks

Fabs commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+// The idea is from an Observable of data and events, compose a stream o React.Components
+const components$ = graphqlObservable(packageRepositoryQuery, schema, {
+  query: fetchRepositories(),
+  filter: searchTerm$
```
Yes. The  data is built lively, so its arguments are also streams.

Fabs commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+import gql from "graphql-tag";
+
+// Streams we get our data from
+import { fetchRepositories } from "#PLUGINS/catalog/src/js/repositories/repositoriesStream";
```
ok. Won't mention marathon or mesos, but mention that it is not an ednpoint.

nLight commented:
I wonder why tests didn't even run

Fabs commented:
@nLight that is actually weird :/

I think on any case that this PR should not be prevented to merge because of tests, since it is just a text file and can't possibly brake tests. 

Maybe @juliangieseke, @DanielMSchmidt, @bstavroulakis or @ahoskins are in a better position to investigate that.

juliangieseke commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+communicate. 
+
+With this pattern, a component will receive down all the data `(data down)` it needs via props,
+and send `events up` so the mediator can perform mutations (represented here as a simple callback). 
```
?

Fabs replied:
That is weird, I will look into it again. I probably have a copy on Grammarly.

juliangieseke commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+With this pattern, a component will receive down all the data `(data down)` it needs via props,
+and send `events up` so the mediator can perform mutations (represented here as a simple callback). 
+The key concept here is that both the data provider and the view component have no idea how each other work.
+Only the caller of `renderComponents` understands how it works.  
```
?

juliangieseke commented on packages/data-service/README.md:
````diff
@@ -0,0 +1,147 @@
+
+```js
+  /*
+   * Receives a stream of React.Component and outputs a single components that 
````
?

juliangieseke commented on packages/data-service/README.md:
````diff
@@ -0,0 +1,147 @@
+  componentFromStream(component$: Observable<React.Component>): React.Component;
+```
+
+Here is an example of how it could be used to render a page for `space shuttles`
````
?

juliangieseke commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+Here is an example of how it could be used to render a page for `space shuttles`
+that shows which shuttles are on space, and which are ready to be launched.
+
+Note that the way we achieve the Mediator pattern is not in an OO fashion, but
```
?

juliangieseke commented on packages/data-service/README.md:
```diff
@@ -0,0 +1,147 @@
+
+## Query Example
+
+To exemplify how you would use the Mediator pattern to get data from a graphql
```
?

Fabs commented:
Sorry @juliangieseke, now we have the latest, latest.

juliangieseke commented:
@Fabs can you force push something again? This weird issue is a Jenkins bug which happens "sometimes" (according to their Bugtracker…)

Fabs commented:
@juliangieseke. I caused it to trigger again :D.
